### PR TITLE
`iceprog' improvements, documentation fixes

### PR DESCRIPTION
**Pull request by rlutz**
This branch improves the command-line interface of the `iceprog' tool:
- Added checks for malformed or invalid command-line options
- More details are given about invocation errors
- Input/output files are opened first so errors can be detected before initializing the hardware
- Exit status indicates whether a software, hardware, or verification error occurred
- Data to be programmed can be read from the standard input
- Fixed option `-R'
- Made `--help' text more detailed
- Added manpage

It also fixes some minor issues in the I/O tile documentation.

### CODE REVIEW

File: iceprog/iceprog.c [hunk 1]
```diff
@@ -407,24 +422,124 @@ int main(int argc, char **argv)
...
+		return EXIT_FAILURE;
+	}
+
+	/* open input/output file in advance
+	   so we can fail before initializing the hardware */
+
+	FILE *f = NULL;
+	long file_size = -1;
+
+	if (test_mode)
```
Comment (rlutz) [inline iceprog/iceprog.c]: It is available from the command line; it's the `-t' option.

A dry run is a conceptually different thing: it doesn't talk to the hardware (`-t' does; that's what it's for) but otherwise goes through the full motions (`-t' doesn't). That said, a dry run mode would probably be a nice thing to have.

File: iceprog/iceprog.c [hunk 2]
```diff
@@ -407,24 +422,124 @@ int main(int argc, char **argv)
...
+		/* nop */;
+	else if (read_mode) {
+		f = (strcmp(filename, "-") == 0) ? stdout
```
Comment (rlutz) [inline iceprog/iceprog.c]: I thought about that when updating the documentation, but given that few people will actually program from stdin (or write the flash contents to stdout) and that this notation is common and just a convenience, I thought it wasn't worth the user's (limited) attention. I see your point, though.

File: iceprog/iceprog.c [hunk 3]
```diff
@@ -438,43 +553,43 @@ int main(int argc, char **argv)
 	if (devstr != NULL) {
 		if (ftdi_usb_open_string(&ftdic, devstr)) {
 			fprintf(stderr, "Can't find iCE FTDI USB device (device string %s).\n", devstr);
-			error();
+			error(2);
```
Comment (rlutz) [inline iceprog/iceprog.c]: I agree that this isn't nice; maybe even an old version of the file has been truncated. The correct way of handling this would be writing to a temporary file whose name is based on the target file (to make sure it's on the same file system) and on success renaming it, replacing the target file. It believe this may be a bit overkill here, but if you think it's worth the additional complexity, I'll implement it.

File: iceprog/iceprog.c (hunk 1 above)
Comment (martinda) [inline iceprog/iceprog.c]: Thanks, I had missed the `-t` option. A dry run mode would be nice, but it is not required. I also just want to capture that as we were discussing this, you thought about documenting that the test mode can be used to clear the reset pin, and that you thought you needed the reverse of this as well.

File: iceprog/iceprog.c (hunk 3 above)
Comment (martinda) [inline iceprog/iceprog.c]: This can be a future improvement. I do not require this.

**Comment by martinda** [comment]
Overall, looks good to me.

**Comment by rlutz** [comment]
*return the reset pin to high (it's active low)

**Comment by cliffordwolf** [comment]
I've now gone over the whole patch. Looks OK for the most part. How well did you test the new code?

For the future: This patch contains a lot of changes that I would describe as "coding style" changes. For example, you moved many curly brackets from the next line to the end of the previous line. I usually put the opening curly bracket in a separate line if the block is a top-level block or the block contains empty lines.

Also: In a function definition (not declaration !) "foobar()" and "foobar(void)" are equivalent. You changed a lot of function declarations to use the void keyword. You also added a lot of "static" declarations, which really does not make much of a difference in the main program.

Same thing for stuff like "break overlong lines". It's not like you add a feature there. You just replace one style guide with another one. It's not like I forgot to press the enter key when I wrote that code. I wrote it this way because I liked it that way.

The usual approach would be do mimic the coding style of the existing code when adding features, not reformat the entire program to match your personal preferences. The next guy will want to revert half of that changes and I will need twice the time to review pull requests because there is so much coding style back-and-forth changing going on that it is hard to spot the relevant changes in the code..

So it would be nice to have pull requests that change "coding style" separate from pull requests that add features.

**Comment by rlutz** [comment]
Ok, so this seems to be a bit of a misunderstanding here. I assumed you were following standard K&R style with tabulator indentation (“kernel coding style”) and just didn't bother to e.g. break the lines.

If this is how you intended the code to be, just skip whatever commits of my pull request you don't like: that's why I kept style changes separate from (and following) everything else.

> In a function definition (not declaration !) "foobar()" and "foobar(void)" are equivalent.

Oops, you're right.  Removed that commit…

> You also added a lot of "static" declarations, which really does not make much of a difference in the main program.

That's true. However, I consider it good style to declare anything as `static` which isn't meant to be visible from other objects. For the main program, that means anything but `main` should be `static`.

As with the other style changes, feel free to skip that commit if you don't agree.

> The usual approach would be do mimic the coding style of the existing code when adding features

As mentioned before, I somehow clicked into the closest coding style I was familiar with and followed it, sorry for that. (To my defense, you do follow K&R style most of the time.) If you want, I can go back and un-break the lines in the new code.

> TL;DR: Please keep large coding style patches separate from small feature patches.

Feature patches: 4 files changed, 378 insertions(+), 138 deletions(-)
Coding style patches: 1 file changed, 108 insertions(+), 91 deletions(-)

> Important question before I merge your changes: How well did you test the new code?

I ran it under various conditions to ensure it works right, but considering the relative simplicity of the code, I didn't run extensive tests.

File: iceprog/iceprog.c (hunk 1 above)
Comment (rlutz) [inline iceprog/iceprog.c]: *return the reset pin to high (it's active low)

**Comment by cliffordwolf** [comment]
I've merged it and made some coding style related changes ontop of your patches (see 3c42bdb).

> I ran it under various conditions to ensure it works right, but considering the relative simplicity of the code, I didn't run extensive tests.

We'll see if someone yells at us for breaking it.. :)
